### PR TITLE
Working Backend and Frontend for CLUE Screen

### PR DESCRIPTION
**Pull request by andreamah**
# Description:

> Before I scare anyone off with this PR, please know that ~25 000 of the added lines of code are from the *.bdf (bitmap defined font) file of the default text typeface (which doesn't need to be reviewed)!

**This initial CLUE PR allows for the following:**
- Backend Python CLUE library support for:
  - adafruit_clue **(only for pixel and display text)**: https://circuitpython.readthedocs.io/projects/clue/en/latest/api.html
  - adafruit_display_text (used by adafruit_clue, but can be used independently also): https://github.com/adafruit/Adafruit_CircuitPython_Display_Text
  - adafruit_bitmap_font (used by adafruit_clue, but can be used independently also): 
https://github.com/adafruit/Adafruit_CircuitPython_Bitmap_Font
  - adafruit_display_shapes: https://github.com/adafruit/Adafruit_CircuitPython_Display_Shapes/
  - adafruit_fancyled (for CPX and CLUE): https://github.com/adafruit/Adafruit_CircuitPython_FancyLED
  - neopixel (for CPX and CLUE): https://github.com/adafruit/Adafruit_CircuitPython_NeoPixel

NOTE 1: adafruit_display_text is present within a tar.gz of the latest release on GitHub, since they seemed to have manually disabled deployment to PyPI. Coincidentally, adafruit_display_text also required a few fixes for off-by-one errors, so that it incorporated in the tar.gz

NOTE 2: adafruit_clue is not downloaded via PyPI, but is a local revised version that is pre-downloaded.

Neopixel (excluding neopixel_write), adafruit_fancyled, adafruit_bitmap_font, and adafruit_display_shapes are all downloaded directly from PyPI upon setup. They are installed as dependencies.

**Directory structure for new folders:**
- clue/
  - the manual adafruit_clue driver and additional tests for some of the libraries above that can work on the clue.
- base_circuitpython/
  - Python implementation of certain Circuitpython (https://circuitpython.readthedocs.io/en/5.0.x/README.html) components. Instead of interacting with hardware, ours interacts with our simulator.

**Frontend-backend communication:**
- Currently, the backend for CLUE only emits to the frontend (no listener on backend yet). Image state is stored in a PIL (Python Imaging Library) Image object and is encoded to base64+sent to frontend whenever is it supposed to update. Work is currently being done to try to make this process more efficient on the backend side.

Frontend Info:
- Currently, the frontend just contains a stripped-down microbit SVG with a png image for a screen. This will later be updated to contain an actual CLUE SVG.

image

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

# Limitations:

Please describe limitations of this PR

# Testing:

Here are some tests that I ran & verified that it looks the same on the device:
- For adafruit_clue (and indirectly, adafruit_display_text)
```
from adafruit_bitmap_font import bitmap_font

from adafruit_clue import clue

clue_data = clue.simple_text_display(title="CLUE Sensor Data!", title_scale=2,) //tested with different scaling

while True:
    clue_data[0].text = "Accelerometer:"
    clue_data[1].text = "Gyro:"
    clue_data[2].text = "Magnetic:"
    clue_data[3].text = "Pressure: {:.3f} hPa".format(100)
    clue_data[4].text = "Altitude: {:.1f} m".format(100)
    clue_data[5].text = "Temperature: {:.1f} C".format(100)
    clue_data[6].text = "Humidity: {:.1f} %".format(100)
    clue_data[7].text = "Proximity: {}".format(100)
    clue_data[8].text = "Gesture: {}".format("flip")
    clue_data[9].text = "Color: R: {} G: {} B: {} C: {}".format(100, 100, 100, 100)
    clue_data[10].text = "Button A: {}".format(False)
    clue_data[11].text = "Button B: {}".format(False)
    clue_data[12].text = "Touch 0: {}".format(False)
    clue_data[13].text = "Touch 1: {}".format(False)
    clue_data[14].text = "Touch 2: {}".format(False)
    clue_data.show()
```
and 
```
from adafruit_clue import clue
while True:
    clue_data.pixel.fill((0, 3, 4))
```
NOTE: it is known that it bugs out if you switch between clue_data.pixel.fill() and clue_data.show(). This is because the pixel is not implemented on the frontend.

- For adafruit_display_shapes
https://github.com/adafruit/Adafruit_CircuitPython_Display_Shapes/blob/master/examples/display_shapes_simpletest.py
- For adafruit_fancyled and neopixel (for CPX and CLUE, diff num_leds for different device tests)
```
import board
import neopixel
import adafruit_fancyled.adafruit_fancyled as fancy

num_leds = 1 // would be 10 for CPX
# Declare a 6-element RGB rainbow palette
palette = [
    fancy.CRGB(1.0, 0.0, 0.5),  # Pink
    fancy.CRGB(0.0, 1.0, 0.0),  # Green
    fancy.CRGB(0.0, 0.0, 1.0),
]  # Blue

pixels = neopixel.NeoPixel(board.NEOPIXEL, num_leds, brightness=1.0, auto_write=False)

offset = 0  # Positional offset into color palette to get it to 'spin'

while True:
    for i in range(num_leds):
        # Load each pixel's color from the palette using an offset, run it
        # through the gamma function, pack RGB value and assign to pixel.
        color = fancy.palette_lookup(palette, offset + i / num_leds)
        color = fancy.gamma_adjust(color, brightness=0.25)
        pixels[i] = color.pack()
    pixels.show()

    offset += 0.02  # Bigger number = faster spin
```

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] My code has been formatted with `npm run format` and passes the checks in `npm run check`
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by markAtMicrosoft — APPROVED** [review]

File: src/base_circuitpython/displayio/group.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+                x += self._anchor_point[0]
+                y += self._anchor_point[1]
+                if self._boundingbox is not None and self.anchored_position is not None:
+                    x += self.anchored_position[0]
+                    y += self.anchored_position[1]
+        except AttributeError:
+            pass
+
+        for idx, elem in enumerate(self.__contents):
```
Comment (vandyliu) [inline src/base_circuitpython/displayio/group.py]: _nit_ don't need enumerate or idx because they're not used

File: gulpfile.js [hunk 2]
```diff
@@ -35,14 +35,18 @@ gulp.task("clean", () => {
...
   "./src/micropython/*.*",
   "./src/micropython/microbit/*.*",
   "./src/micropython/microbit/!(test)/**/*",
   "./src/*.py",
   "./src/common/*.py",
   "./src/dev-requirements.txt",
   "./src/requirements.txt",
-  "./src/templates/*.*"
+  "./src/templates/*.*",
+  "./src/*.sh"
```
Comment (vandyliu) [inline gulpfile.js]: Do we still have shell scripts?

File: locales/en/package.i18n.json [hunk 3]
```diff
@@ -1,21 +1,22 @@
...
-    "deviceSimulatorExpressExtension.commands.microbit.deployToDevice": "[micro:bit] Deploy to Device",
-    "deviceSimulatorExpressExtension.commands.microbit.openSimulator": "[micro:bit] Open Simulator",
-    "deviceSimulatorExpressExtension.commands.microbit.newFile": "[micro:bit] New File",
-    "deviceSimulatorExpressExtension.configuration.title": "Device Simulator Express configuration",
-    "deviceSimulatorExpressExtension.configuration.properties.configEnvOnChange": "When you change the Python interpreter, the Device Simulator Express will automatically configure itself for the required dependencies.",
-    "deviceSimulatorExpressExtension.configuration.properties.debuggerPort": "The port the Server will listen on for communication with the debugger.",
-    "deviceSimulatorExpressExtension.configuration.properties.dependencyChecker": "Whether or not to ask if we can download dependencies. If unchecked, the extension will default to never download dependencies, except when automatically creating a virtual environment in the extension files."
-}
+  "deviceSimulatorExpressExtension.commands.common.installDependencies": "Install Extension Dependencies",
```
Comment (vandyliu) [inline locales/en/package.i18n.json]: Can we not change the line spacing because it has no changes ;~; ?

File: .vscode/settings.json [hunk 4]
```diff
@@ -8,4 +8,4 @@
     },
     // Turn off tsc task auto detection since we have the necessary tasks as npm scripts
     "typescript.tsc.autoDetect": "off"
-}
+}
```
Comment (vandyliu) [inline .vscode/settings.json]: Can we not include this because it has no changes ;~; ?

File: package.nls.json [hunk 5]
```diff
@@ -1,21 +1,22 @@
...
-    "deviceSimulatorExpressExtension.commands.microbit.deployToDevice": "[micro:bit] Deploy to Device",
-    "deviceSimulatorExpressExtension.commands.microbit.openSimulator": "[micro:bit] Open Simulator",
-    "deviceSimulatorExpressExtension.commands.microbit.newFile": "[micro:bit] New File",
-    "deviceSimulatorExpressExtension.configuration.title": "Device Simulator Express configuration",
-    "deviceSimulatorExpressExtension.configuration.properties.configEnvOnChange": "When you change the Python interpreter, the Device Simulator Express will automatically configure itself for the required dependencies.",
-    "deviceSimulatorExpressExtension.configuration.properties.debuggerPort": "The port the Server will listen on for communication with the debugger.",
-    "deviceSimulatorExpressExtension.configuration.properties.dependencyChecker": "Whether or not to ask for dependency downloads. If unchecked, the extension will default to never download dependencies, except when automatically creating a virtual environment in the extension files."
-}
+  "deviceSimulatorExpressExtension.commands.common.installDependencies": "Install Extension Dependencies",
```
Comment (vandyliu) [inline package.nls.json]: Can we not change the line spacing because it has no changes ;~; ?

File: src/view/components/clue/ClueImage.tsx [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+//             const ledElement = ledRefs[j][i].current;
+//             if (ledElement) {
+//                 setupLed(ledElement, leds[i][j]);
+//             }
+//         }
+//     }
+// };
+// const setupLed = (ledElement: SVGRectElement, brightness: number) => {
+//     ledElement.style.opacity = (brightness / 10).toString();
+// };
```
Comment (vandyliu) [inline src/view/components/clue/ClueImage.tsx]: Let's remove this?

File: src/view/components/clue/ClueSimulator.tsx [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    AB_BUTTONS_KEYS,
+    WEBVIEW_MESSAGES,
+} from "../../constants";
+import PlayLogo from "../../svgs/play_svg";
+import StopLogo from "../../svgs/stop_svg";
+import { sendMessage } from "../../utils/MessageUtils";
+import ActionBar from "../simulator/ActionBar";
+import { BUTTONS_KEYS, ClueImage } from "./ClueImage";
+
+// import * as fs from "fs";
```
Comment (vandyliu) [inline src/view/components/clue/ClueSimulator.tsx]: Should we remove commented out imports? Same as line 4.

File: src/process_user_code.py [hunk 8]
```diff
@@ -44,14 +50,16 @@
...
+
 
 # Handle User Inputs Thread
 class UserInput(threading.Thread):
     def __init__(self):
         threading.Thread.__init__(self)
 
     def run(self):
-        device_dict = {CPX: cpx, MICROBIT: mb}
+        device_dict = {CPX: cpx, MICROBIT: mb, "CLUE": clue}
```
Comment (vandyliu) [inline src/process_user_code.py]: Can we use clue constant like the others?

File: src/extension.ts [hunk 9]
```diff
@@ -715,7 +748,9 @@ export async function activate(context: vscode.ExtensionContext) {
             // Data received from Python process
             deviceProcess.stdout.on("data", data => {
                 dataFromTheProcess = data.toString();
+                console.log("here 1");
                 console.log(`Device output = ${dataFromTheProcess}`);
+                console.log("here 2");
```
Comment (vandyliu) [inline src/extension.ts]: Can we remove these "here" console.logs?

File: src/extension.ts [hunk 10]
```diff
@@ -338,6 +344,18 @@ export async function activate(context: vscode.ExtensionContext) {
...
+    const clueOpenSimulator: vscode.Disposable = vscode.commands.registerCommand(
+        "deviceSimulatorExpress.clue.openSimulator",
+        () => {
+            telemetryAI.trackFeatureUsage(
+                TelemetryEventName.MICROBIT_COMMAND_OPEN_SIMULATOR
+            );
+            telemetryAI.runWithLatencyMeasure(
+                openClueWebview,
+                TelemetryEventName.MICROBIT_PERFORMANCE_OPEN_SIMULATOR
+            );
```
Comment (vandyliu) [inline src/extension.ts]: I think we should just make it run openClueWebview() for now and have a task to add telemetry later instead of using microbit telemetry.

File: src/extension.ts [hunk 11]
```diff
@@ -441,6 +465,19 @@ export async function activate(context: vscode.ExtensionContext) {
...
+        "deviceSimulatorExpress.clue.newFile",
+        () => {
+            telemetryAI.trackFeatureUsage(
+                TelemetryEventName.MICROBIT_COMMAND_NEW_FILE
+            );
+            telemetryAI.runWithLatencyMeasure(
+                openClueTemplateFile,
+
+                TelemetryEventName.MICROBIT_PERFORMANCE_NEW_FILE
+            );
```
Comment (vandyliu) [inline src/extension.ts]: I think we should just make it run openClueTemplateFile() for now and have a task to add telemetry later instead of using microbit telemetry.

File: src/clue/test/test_helpers.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    def __test_image_equality(self, image_1, image_2):
+        for i in range(CONSTANTS.SCREEN_HEIGHT_WIDTH):
+            for j in range(CONSTANTS.SCREEN_HEIGHT_WIDTH):
+                pixel_1 = image_1[j, i]
+                pixel_2 = image_2[j, i]
+                # if not isinstance(pixel_1, tuple):
+                #     pixel_1 = self.__convert_hex_to_rgb(pixel_1)
+
+                # if not isinstance(pixel_2, tuple):
+                #     pixel_2 = self.__convert_hex_to_rgb(pixel_2)
```
Comment (vandyliu) [inline src/clue/test/test_helpers.py]: We should remove these comments if they aren't necessary

File: src/clue/test/constants.py [hunk 13]
```diff
@@ -0,0 +1,2 @@
+IMG_DIR_NAME = "img"
+SCREEN_HEIGHT_WIDTH = 240
```
Comment (vandyliu) [inline src/clue/test/constants.py]: In places, where we used this constants file, can we use the one that the clue uses?

File: src/view/components/clue/Clue.tsx [hunk 14]
```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as React from "react";
+import { MICROBIT_TOOLBAR_ID } from "../toolbar/SensorModalUtils";
```
Comment (vandyliu) [inline src/view/components/clue/Clue.tsx]: Should we remove the Toolbar from this PR for now? I know we'll fix it soon but having it here makes it kind of messy and we could forget some things.  We can just have a comment `// Todo: toolbar` (pt 1)

File: src/view/components/clue/Clue.tsx [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+        }
+    };
+    render() {
+        return (
+            <React.Fragment>
+                <ClueSimulator />
+                <ToolBar
+                    buttonList={MICROBIT_TOOLBAR_BUTTONS}
+                    onUpdateSensor={this.updateSensor}
+                    sensorValues={this.state.sensors}
```
Comment (vandyliu) [inline src/view/components/clue/Clue.tsx]: (pt 2) so just put // Todo: toolbar

File: src/view/components/clue/Clue.tsx [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    },
+    {
+        image: TOOLBAR_SVG.LIGHT_SVG,
+        label: MICROBIT_TOOLBAR_ID.LIGHT,
+    },
+    {
+        image: TOOLBAR_SVG.MOTION_SVG,
+        label: MICROBIT_TOOLBAR_ID.ACCELEROMETER,
+    },
+];
```
Comment (vandyliu) [inline src/view/components/clue/Clue.tsx]: (pt 3)

**Review by vandyliu — CHANGES_REQUESTED** [review]
Can't believe I just read all of that but it was worth it! Excellent job!!!!!!!!!!!!!!!!!!!!!!! 👯 💃 👯‍♂ 🕺

File: locales/en/package.i18n.json (hunk 3 above)
Comment (xnkevinnguyen) [reply locales/en/package.i18n.json]: I'm not sure why, everytime there's a change here, it changes the whole file. There were added commands for clue

Comment (andreamah) [reply locales/en/package.i18n.json]: changed the spacing, will commit soon

File: gulpfile.js (hunk 2 above)
Comment (andreamah) [reply gulpfile.js]: nope, removing the line!

File: .vscode/settings.json (hunk 4 above)
Comment (andreamah) [reply .vscode/settings.json]: I'm not sure why there's a difference there.. do you know what it is?

Comment (andreamah) [reply .vscode/settings.json]: oops, it's a newline... I'll look into it!

**Review by vandyliu — APPROVED** [review]
Looks good!!! :D 
Excited to merge it.

File: src/clue/adafruit_clue.py [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+            self._y = 3
+
+        self._lines = []
+        for num in range(1):
+            self._lines.append(self.add_text_line(color=colors[num % len(colors)]))
+
+    def __getitem__(self, item):
+        """Fetch the Nth text line Group"""
+        if len(self._lines) - 1 < item:
+            for _ in range(item - (len(self._lines) - 1)):
```
Comment (isadorasophia) [inline src/clue/adafruit_clue.py:146]: Document that we are initializing up to the item-th line in this loop?

File: src/clue/adafruit_clue.py [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+        else:
+            self._y = 3
+
+        self._lines = []
+        for num in range(1):
+            self._lines.append(self.add_text_line(color=colors[num % len(colors)]))
+
+    def __getitem__(self, item):
+        """Fetch the Nth text line Group"""
+        if len(self._lines) - 1 < item:
```
Comment (isadorasophia) [inline src/clue/adafruit_clue.py:145]: Is there any limit for `item` that we should worry about?

File: src/clue/adafruit_clue.py [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+        return text_label
+
+    def show(self):
+        """Call show() to display the data list."""
+        self.text_group.draw(show=True)
+        # https://stackoverflow.com/questions/31826335/how-to-convert-pil-image-image-object-to-base64-string
+
+    def show_terminal(self):
+        """Revert to terminalio screen."""
+        # self._display.show(None)
```
Comment (isadorasophia) [inline src/clue/adafruit_clue.py]: Is this a TODO?

File: src/base_circuitpython/displayio/tile_grid.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+        curr_colour,
+        x_offset,
+        y_offset,
+        scale,
+        x_max,
+        y_max,
+        this_bmp_img,
+    ):
+
+        for new_y in range(y_offset, y_max):
```
Comment (isadorasophia) [inline src/base_circuitpython/displayio/tile_grid.py:143]: Is this the drawing logic which we are looking to improve?

File: src/base_circuitpython/neopixel_write.py [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+    """Write buf out on the given DigitalInOut."""
+
+    if len(tuple(buf)) > 0:
+        # if we are explicitly given
+        # the clue pin, that means that
+        # the clue is definitely the active device
+        # because the constructor for the
+        # clue is what calls neopixel
+        # with the clue pin argument
+        if gpio.pin != CONSTANTS.CLUE_PIN:
```
Comment (isadorasophia) [inline src/base_circuitpython/neopixel_write.py:26]: Do we expect any pin for cpx?

**Review by isadorasophia — APPROVED** [review]
Super cool! Really nice job on the documentation too 😁

File: src/clue/adafruit_clue.py (hunk 19 above)
Comment (andreamah) [reply src/clue/adafruit_clue.py]: yes! I'll add TODO to it!

File: src/clue/adafruit_clue.py (hunk 18 above)
Comment (andreamah) [reply src/clue/adafruit_clue.py:145]: that's mainly code from adafruit... sorry, I didn't separate the adafruit code and manual code very well!! :(

File: src/base_circuitpython/displayio/tile_grid.py (hunk 20 above)
Comment (andreamah) [reply src/base_circuitpython/displayio/tile_grid.py:143]: Yup! Pretty much everything following the "draw()" method in TileGrid

File: src/base_circuitpython/neopixel_write.py (hunk 21 above)
Comment (andreamah) [reply src/base_circuitpython/neopixel_write.py:26]: Ah, I wasn't very clear with how I implemented this!

So in real life, both the CLUE and CPX use the neopixel library; however, when they were developing the CPX, they created pixel.py, which is pretty much an edited version of the neopixel library. 

In my implementation, my clue library uses the neopixel library, which means that all streams of neopixel data go through my neopixel_write function. They can also directly import the neopixel library and interact with it directly. These are both either directly or indirectly sending data through neopixel.py

At the same time, on the CPX, simulator users can now also directly import neopixel (new with the neopixel library) or uses the field in the cpx library (how we've always done it, and controlled in pixel.py). Therefore, pixel info from the CPX can come from neopixel.py or pixel.py.

Therefore, when using the neopixel library, I have it sending two messages: one to the CPX frontend and another to the CLUE backend. This can be a little inefficient, especially if the neopixel was called by the CLUE library in the first place! Therefore, I have my custom clue library instantiate the neopixel library using the CLUE_PIN constant, in order to tell it to never send its light values to the CPX, and that it only needs to send to the clue. If the user manually imported neopixel to use with the CLUE, the neopixel library wouldn't have CLUE_PIN for its gpio pin and it would send to both devices. 

**TLDR; the CLUE_PIN is only used when the clue library instantiates a neopixel instance. The CPX never instantiates a neopixel instance because it was built using its own neopixel library. I am hoping that we can integrate the actual neopixel library with the CPX sometime in the future, but the task seems a little out of bounds for this PR**


Sorry for the long read 😥😥

File: src/clue/adafruit_clue.py (hunk 17 above)
Comment (andreamah) [reply src/clue/adafruit_clue.py:146]: Same as the other comment! This is code by Adafruit themselves :'''''(

File: src/base_circuitpython/displayio/tile_grid.py (hunk 20 above)
Comment (isadorasophia) [reply src/base_circuitpython/displayio/tile_grid.py:143]: I see, it looks like we are drawing each pixel every time 😮 (you probably tweaked it by now, I was just curious haha)

File: src/clue/adafruit_clue.py (hunk 17 above)
Comment (isadorasophia) [reply src/clue/adafruit_clue.py:146]: Got it, no worries! I think the comment might still be helpful though (since you are the code "owner" now, so any git blames will get back to you 😁).

File: src/base_circuitpython/neopixel_write.py (hunk 21 above)
Comment (isadorasophia) [reply src/base_circuitpython/neopixel_write.py:26]: No worries, thanks for going through this! Just so we can keep track of this, I'd recommend adding a comment or creating a work item/issue on GitHub about that. This should be good for this PR though.
